Basic create tool requires (#2605)
conan-io/docs: tutorial/creating_packages/other_types_of_packages/tool_requires_packages.rst

======================

In the ":ref:`Using build tools as Conan packages <consuming_packages_tool_requires>`" section we learned how to use
a tool require to build (or help building) our project or Conan package.
In this section we are going to learn how to create a recipe for a tool require.

Please, first clone the sources to recreate this project. You can find them in the
`examples2.0 repository <https://github.com/conan-io/examples2>`_ on GitHub:

.. code-block:: bash

    $ git clone https://github.com/conan-io/examples2.git
    $ cd examples2/tutorial/creating_packages/other_packages/tool_requires

A simple tool require recipe
----------------------------

Let's see a recipe that even being very simple, covers most of the ``tool-require`` use-cases:

.. code-block:: python
    :caption: conanfile.py

    import os
    import stat
    from conan import ConanFile
    from conan.tools.files import save

    class my_tool(ConanFile):
        name = "my_tool"
        version = "1.0"
        package_type = "application"

        settings = "os", "arch"

        def package(self):
            extension = "sh" if self.settings_build.os != "Windows" else "bat"

>             app_path = os.path.join(self.package_folder, "bin", "say_hello.{}".format(extension))

Reviewer: I think I would prefer to see a more close to reality example here, with an executable... maybe some tool that verifies something in the consumer build? Because this could be closer to the typicall question that sometines users ask on "how to package all their build scripts and use them in another build?" we may put in a separate example

Reviewer 2: I also want to see the case of the executable, I suggested that above, as the possible following example.
But I don't dislike this one either, I think it is a super common use case. We know tons of users that are packaging scripts, cmake utilities and other stuff in ``build_requires``. Maybe even more common that packaging executable tools?

Reviewer: As I see it, I would explain the case of an executable in the main tutorial history (something inspired in tools like cmake, meson, ninja, nasm... ) and then the case of packaging build scripts in an example. If we want to show the case of the build scripts here in the main tutorial then I would make it a bit closer to the usual case exporting the scripts, not generating them... and make the title more close to: "How to package my build scripts to use in other recipes" that is the question that people will search for.

Reviewer 2: Yes, that makes sense too, the case of using exported scripts, I agree, I already commented that.

I'd probably keep the "package my build script" here in the main tutorial line, not in an example, but I am fine either way.

Author: This is not trying to show how to package "build shell scripts" but the concept of packaging any application to be run in build. But as it is confusing you I'll try to change it. Before you suggest, I'll not use the ``cmake_exe`` template because the readers will learn nothing.

            save(self, app_path, 'echo "Hello!!!"')
            os.chmod(app_path, os.stat(app_path).st_mode | stat.S_IEXEC)

        def package_info(self):
            self.buildenv_info.define("MYVAR", "23")

There are few relevant things in this recipe:

1. It declares ``package_type = "application"``, this is optional but convenient, it will indicate conan that the current
   package doesn't contain headers or libraries to be linked. The consumers will know that this package is an application.

2. The ``package()`` method is packaging the executables into the ``bin/`` folder, that is declared by default as a bindir:
   ``self.cpp_info.bindirs = ["bin"]``.

3. We are using ``self.buildenv_info`` to define environment variables that will be available in the consumer.

As the application is a simple shell script we are granting execution permissions to the script, this is usually not
needed if you are building a real executable.

Let's create a binary package for the tool_require:

.. code-block:: bash

    $ conan create .
    ...
    my_tool/1.0: Calling package()
    my_tool/1.0 package(): Packaged 1 '.sh' file: say_hello.sh

>     my_tool/1.0 package(): Packaged 1 '.bat' file: say_hello.bat

Reviewer 2: This is one or the other line, right?

    my_tool/1.0: Full package reference: my_tool/1.0#a05ef273264042c2082499a3061fb8df:82339...2e7c91ab18a1#dd38aae...37c60cbc

We can create a consumer recipe to test if we can run the ``say_hello`` application of the ``tool_require``.

.. code-block:: python
    :caption: consumer.py

    from conan import ConanFile

    class MyConsumer(ConanFile):
        name = "my_library"
        version = "1.0"
        settings = "os", "arch", "compiler", "build_type"
        tool_requires = "my_tool/1.0"

        def build(self):
            if self.settings_build.os != "Windows"
                self.run("say_hello.sh")
                self.run("echo MYVAR=$MYVAR")
            else:
                self.run("say_hello.bat")

>                 self.run("echo MYVAR=%MYVAR%")

Reviewer 2: ```suggestion
                self.run("set MYVAR")
```
To match the example

    consumer.py (my_library/1.0): Aggregating env generators
    consumer.py (my_library/1.0): Calling build()
    consumer.py (my_library/1.0): RUN: say_hello.sh
    Capturing current environment in ../deactivate_conanbuildenv-release-x86_64.sh
    Configuring environment variables
    Hello!!!
    ...
    MYVAR=23

We can see the ``Hello!!!`` message, and the "23" value assigned to ``MYVAR`` but, why are these automatically available?

- The generators ``VirtualBuildEnv`` and ``VirtualRunEnv`` are automatically used.
- The ``VirtualRunEnv`` is reading the ``tool-requires`` and is creating a launcher like ``conanbuildenv-release-x86_64.sh`` appending
  all ``cpp_info.bindirs`` to the ``PATH``, all the ``cpp_info.libdirs`` to the ``LD_LIBRARY_PATH`` environment variable and
  declaring each variable of ``self.buildenv_info``.
- Every time conan executes the ``self.run``, by default, activates the ``conanbuild.sh`` file before calling any command.
  The ``conanbuild.sh`` is including the ``conanbuildenv-release-x86_64.sh``, so the application is in the PATH
  and the enviornment variable "MYVAR" has the value declared in the tool-require.

> More complex recipes

Author: What else do we want to explain?

Reviewer 2: Good question, maybe the case of using a real .exe? So it requires settings, and the build-host context becomes more evident regarding ``tool_requires``

Reviewer 2: Those could be ReadMore examples, not necessarily here in main line of the tutorial

With the previous recipe, if we call :command:`conan create` with different setting like different compiler versions, we will get
different binary packages with a different ``package ID``. This might be convenient to, for example, keep better traceability of
our tools. In this case, the <MISSING PAGE> compatibility.py plugin can help to locate the best matching binary in case Conan doesn't find the
binary for our specific compiler version.

But in some cases we might want to just generate a binary taking into account only the ``os``, ``arch`` or at most
adding the ``build_type`` to know if the application is built for Debug or Release. We can add a ``package_id()`` method
to remove them:

.. code-block:: python
    :caption: conanfile.py

    import os
    from conan import ConanFile
    from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout
    from conan.tools.files import copy

    class secure_scannerRecipe(ConanFile):
        name = "secure_scanner"
        version = "1.0"
        settings = "os", "compiler", "build_type", "arch"
        ...

        def package_id(self):
            del self.info.settings.compiler
            del self.info.settings.build_type

> So, if we call :command:`conan create` with different ``build_type`` we will get exactly the same binary.

Reviewer 2: ```suggestion
So, if we call :command:`conan create` with different ``build_type`` we will get exactly the same ``package_id``.
```

.. code-block:: bash

    $ conan create .
    ...
    Package '82339cc4d6db7990c1830d274cd12e7c91ab18a1' created

    $ conan create . -s build_type=Debug
    ...
    Package '82339cc4d6db7990c1830d274cd12e7c91ab18a1' created

> We got the same binary package.

Reviewer 2: ```suggestion
We got the same binary ``package_id``. The second ``conan create . -s build_type=Debug`` created and overwrote (created a newer package revision) of the previous Release binary, because they have the same ``package_id`` identifier. It is typical to create only the ``Release`` one, and if for any reason managing both Debug and Release binaries is intended, then the approach would be not removing the ``del self.info.settings.build_type``
```
